_posts/2020-09-30-intro-to-hive-connector.md

```diff
@@ -0,0 +1,111 @@
+---
+layout: post
+title:  "A gentle introduction to the Hive connector"
+author: Brian Olsen
+excerpt_separator: <!--more-->
+---
+One of the most confusing aspects when starting Presto is the Hive connector.
+
+Typically, you seek out the use of Presto when you experience an intensely slow
+query turnaround from your existing Hadoop or Hive infrastructure. In fact, the
+genesis of Presto came about due to these slow Hive query conditions at Facebook
+back in 2012. So when you learn that Presto has a Hive connector, it can be
+rather confusing since you moved to Presto to circumvent the slowness of your
+current Hive cluster. Another common source of confusion is when you want to 
+query your data from your S3 compatible object storage such as AWS S3, MinIO, 
+and Google Cloud Storage. This too uses the Hive connector and if that confuses
+you, this post will explain why.
+
+To understand the origins and inner workings of Presto's Hive connector, you
+first need to know a few high level components of the Hive architecture.
+
+![](/assets/blog/intro-to-hive-connector/hive.png)
+
+
+<!--more-->
+
+For this conversation, you can simplify the Hive architecture to three elements;
+_the runtime_, _the metastore_, and _the storage_. _The runtime_ contains the
+logic of the query engine that translates the SQL-esque Hive Query Language
+(HQL) into MapReduce jobs that run over files stored in the filesystem.
+_The storage_ component is simply that, it stores files in various formats and
+index structures to recall these files. The file formats can be anything as
+simple as JSON and CSV to more complex files such as columnar formats like ORC
+and Parquet. Traditionally, Hive ran on top of the Hadoop Distributed Filesystem
+(HDFS). As cloud-based options became more prevalent, object storage that used 
+Amazon's S3 interface become more and more common and Hive expanded to include
+storing files in Amazon S3, Azure Blob Storage, Google Cloud Storage (GCS), and
+others. In order for Hive to process these files, it must have a mapping from
+SQL tables in _the runtime_ to files and directories in _the storage_ component.
+To accomplish this, Hive uses _the metastore_ to keep the metadata about the
+files such as table columns, file locations, file formats, etc... 
+
+In the early days of Big Data systems, many expected query turnaround took a 
+long time because the data that ETL operated on was big. The primary goal in
+early iterations of these systems was simply throughput over large volumes of
+data while maintaining fault-tolerance. Over time, more businesses wanted to run
+fast interactive queries over their Big Data sets instead of running jobs that
+took hours and produced possibly undesirable results. Upon Presto's creation, 
+there were 300 petabytes of data and all the metadata for this data already
+existed. The data in storage was cumbersome to move, and the data in the
+metastore would have taken a long time to repopulate in another format. The
```

Comment:
This whole paragraph should be changed to present tense and not using "would". and such .. lots of people still have these problems right now